code/modules/security/genpop.dm

```diff
@@ -0,0 +1,650 @@
+		state = TURNSTILE_CIRCUIT_EXPOSED
+		return
+	if(item.tool_behaviour == TOOL_CROWBAR && panel_open && circuit != null)
+		to_chat(user, "<span class='notice'>You start tearing out the circuitry...")
+		if(do_after(user, 3 SECONDS))
+			circuit.forceMove(loc)
+			circuit = null
+			state = TURNSTILE_SHELL
+		return
+	. = ..()
+	if(machine_stat & BROKEN)
+		try_shock(user)
+
+//Shock attack if something is thrown at it if we're broken
+/obj/machinery/turnstile/hitby(atom/movable/AM, skipcatch, hitpush, blocked, datum/thrownthing/throwingdatum)
+	if(isobj(AM))
+		if(prob(50) && BROKEN)
+			var/obj/O = AM
+			if(O.throwforce != 0)//don't want to let people spam tesla bolts, this way it will break after time
+				playsound(src, 'sound/magic/lightningshock.ogg', 100, 1, extrarange = 5)
+				tesla_zap(src, 3, 8000, TESLA_MOB_DAMAGE | TESLA_OBJ_DAMAGE | TESLA_MOB_STUN | TESLA_ALLOW_DUPLICATES) // Around 20 damage for humans
+	return ..()
+
+/obj/machinery/turnstile/welder_act(mob/living/user, obj/item/I)
+	//Shamelessly copied airlock code
+	. = TRUE
+	if(circuit == null && user.a_intent == INTENT_HARM)
+		var/obj/item/weldingtool/W = I
+		if(W.welding)
+			if(W.use_tool(src, user, 40, volume=50))
+				to_chat(user, "<span class='notice'>You start slicing off the bars of the [src]")
+				new /obj/item/stack/rods/ten(get_turf(src))
+				qdel(src)
+	if(!I.tool_start_check(user, amount=0))
+		return
+	if(obj_integrity >= max_integrity)
+		to_chat(user, "<span class='notice'>The turnstile doesn't need repairing.</span>")
+		return
+	user.visible_message("[user] is welding the turnstile.", \
+				"<span class='notice'>You begin repairing the turnstile...</span>", \
+				"<span class='italics'>You hear welding.</span>")
+	if(I.use_tool(src, user, 40, volume=50))
+		obj_integrity = max_integrity
+		set_machine_stat(machine_stat & ~BROKEN)
+		user.visible_message("[user.name] has repaired [src].", \
+							"<span class='notice'>You finish repairing the turnstile.</span>")
+		update_icon()
+		return
+
+/obj/machinery/turnstile/CanAllowThrough(atom/movable/mover, turf/target)
+	. = ..()
+	if(. == TRUE)
+		return TRUE //Allow certain things declared with pass_flags_self through wihout side effects
+	if(machine_stat & BROKEN)
+		return FALSE
+	if(get_dir(loc, target) == dir) //Always let people through in one direction. Not used at the moment.
+		return TRUE
+	var/allowed = allowed(mover)
+	//Everyone with access can drag you out. Prisoners can't drag each other out.
+	if(istype(mover, /obj/vehicle/ridden) && mover.buckled_mobs)
+		playsound(loc, 'sound/machines/buzz-sigh.ogg', 50)
+		say("ERROR. For security reasons, wheelchairs and other ridden devices are not allowed through the turnstile.")
+		return FALSE
+	if(!allowed && mover.pulledby && ishuman(mover.pulledby))
+		var/mob/living/carbon/human/H = mover.pulledby
+		if(istype(H.wear_id, /obj/item/card/id/prisoner) || istype(H.get_active_held_item(), /obj/item/card/id/prisoner))
```

Comment:
It might be better to check for security access instead of type here

Reply:
Unless I'm reading the code wrong. This here prevents prisoners from being dragged by another prisoner through the turnstile. Prisoners are allowed through the turnstile once their time is served, so what would happen without this check is prisoners being able to commit evasion simplly by being pulled through a prisoner that had been released.
I deemed it would be better if that type of evasion was only possible through a physical exchange of the ID rather than a cheap method like this that offers little to not counterplay for security.

Checking for type prevents that, checking for security access doesn't prevent this flaw because prisoners that are released don't get security access.